### ISSUE TITLE
Best way to update each agent after the other

### ISSUE THREAD

**Issue by BenediktAllendorf** [JuliaDynamics/Agents.jl#418]
Sorry to hijack the issue tracker; I wasn't sure where else to post my question.

I'm wondering what the best way to update each agent after the other is, and I'm looking at how the opinion spread example does it [1].

In general, there seem to be two ways: within agent_step! (as in the example) or with model_step! where one would loop through all agents, save the changed value to some temporary variable, then loop again and update the agents' values.

Is there a "best practice" to this?

I can see how the example avoids a second loop and reallocating a new temporary variable, but therefore the agent-struct gets bigger. And if you have multiple attributes to change, the code gets ugly (as each attribute would be an additional prev_-attribute). Is the chosen way more performant, or is this only about coding style?

[1] https://juliadynamics.github.io/Agents.jl/stable/examples/opinion_spread/

**Comment by Libbum** [comment]
In terms of best-practice: no, there isn't one. The agent/model step separation is entirely for convenience for many models that don't have the tight requirements that you have in your situation. Using model_step! by itself is the way to go if you need such an update. We describe the general situation in [Advanced Stepping](https://juliadynamics.github.io/Agents.jl/stable/tutorial/#Advanced-stepping) in the documentation.

There's been a long discussion about what's best here (#331), and the current code style was the way we've settled on for now&mdash;nothing is really tied to performance here, more-so style and convenience.

If you want to post an example of your problem we can give you some pointers.

**Comment by BenediktAllendorf** [comment]
Thank you for pointing them out! My "problem" is within all that discussion and I guess right now it's more a personal choice how you implement it. To be honest, I don't like the solution (in the opinion or HK example) yet. They take something that is conceptually simple (synchronous updates) and make it kind of complicated.

But this seems to be more a general Julia/programming problem and not Agents.jl specific (which I could have noticed if I re-read the new documentation - when I did the last time, this wasn't included or obvious yet). I'm brainstorming a few ideas on how to improve this (and maybe will even benchmark them). I'll let you know! :)

**Comment by Libbum** [comment]
Sounds great! 

The discussion in those threads was more about the model/agent step! divide, rather than figuring out if the current synchronous update syntax is the best. So for sure&mdash;would be great to see what you come up with!

**Comment by AayushSabharwal** [comment]
I actually have an idea on this front. An alternative solution could be queueing up "actions" in each `agent_step!`, where each action identifies the new value for a field of an agent. For example, an action can be of the form `(id, symbol, value)`. Then, after all `agent_step!`s, for each such action, agent with ID `id` has it's field `symbol` updated to `value`. 

Updating complex fields through this method could be difficult though, for example if a field is an `Array` or another struct. Additionally, there might be some performance implications since `value` doesn't have a fixed type.

Just throwing the idea out there, if it might evolve into something better. Personally, I think I prefer the current method of doing things. It requires a couple extra fields, but it fits well with the existing syntax, doesn't require any extra work on the library's part and is probably as memory efficient as possible. I can see how this can become an issue with a large number of fields though.

**Comment by BenediktAllendorf** [comment]
Yes, precisely what I was thinking about.

I did some hacking and got to this version:

```julia
test = ones(Int, 4)

@atonce for x in 1:4
    @pending test[x] = sum(test)
end
```

Without the macros, `test` would be: `4 7 13 25`, and with them, it's `4 4 4 4`. 


How it works:
```julia
let laterstorage = Dict()
    global function setfieldlater(a, b, c)
        laterstorage[(a, b)] = c
    end

    global function getlaterstorage()
        return laterstorage
    end

    global function setlaterstorage(stor::AbstractDict)
        laterstorage = stor
    end
    
    global function clearlaterstorage()
        laterstorage = Dict()
    end

    function set(a, b, c)
        setproperty!(a, b, c)
    end
    
    function set(a::Union{AbstractArray, AbstractDict}, b, c)
        setindex!(a, c, b)
    end

    global function laterisnow()
        for ((a,b), c) in laterstorage
            set(a, b, c)
        end
    
        clearlaterstorage()
    end
end

macro pending(expr)
    if expr.head != :(=)
        throw(DomainError(expr.head, "You can only use @pending with '='!"))
    end
    
    if length(expr.args) != 2
        throw(DomainError(expr, "You can't use @pending like that!"))
    end

    if !isa(expr.args[1], Expr)
        throw(DomainError(expr, "@pending only works with references!"))
    end
    
    Expr(:call, setfieldlater, esc(expr.args[1].args[1]), esc(expr.args[1].args[2]), esc(expr.args[2]))
end

macro atonce(body)
    quote
        stor = getlaterstorage()
        setlaterstorage(Dict())
        $(esc(body))
        laterisnow()
        setlaterstorage(stor)
    end
end
```

I guess this is slower than other alternatives (because it has to allocate more and without fixed types), but I like how you don't have to think about implementing the synchronous update yourself.

**Comment by Libbum** [comment]
Oh, I see. Perhaps I miss-understood your initial question actually. I follow your requirements a little better now, but don't (yet) see this solution as necessary.

Would not
```julia
test = ones(Int, 4)
map(t->sum(test), test)
```
yield the same results? At least it does for this test case. Maybe if we could take a look at the Opinion Spread implementation in a little more depth together (another contributor wrote it). See how your solution would make the process easier, and I'll take a shot at writing it without the temp variables etc.

**Comment by BenediktAllendorf** [comment]
Yeah, sure, the example was too simple to make the point. It really only makes sense if the updates depend on each other, like this:

```julia
test = [1.0, 2.0, 1.0, 2.0, 1.0]

@atonce for x in 2:4
    @pending test[x] = mean(test[x-1:x+1])
end
```

Then it's `1.0 1.33 1.66 1.33 1.0` vs `1.0 1.33 1.44 1.48 1.0`.


Taking the HK example, instead of
```julia
mutable struct HKAgent <: AbstractAgent
    id::Int
    old_opinion::Float64
    new_opinion::Float64
    # I removed `previous_opinon` because that's unrelated
end

function agent_step!(agent, model)
    agent.new_opinion = mean(boundfilter(agent, model))
end

function model_step!(model)
    for a in allagents(model)
        a.old_opinion = a.new_opinion
    end
end
```
you could have
```julia
mutable struct HKAgent <: AbstractAgent
    id::Int
    opinion::Float64
end

function model_step!(model)
    @atonce for a in allagents(model)
        @pending a.opinion = mean(boundfilter(agent, model))
    end
end
```

**Comment by Libbum** [comment]
That's a good example to start with. Things we'd need to consider is some type safety, and performance questions at the very least. If we can do away with the need for macros, that would be good; but maybe it's not possible. The names may need some tweaking, but that's of minor concern at the moment.

With all that said, I don't foresee any reason why we couldn't get performance parity with the current implementation, but with the added bonus of simpler agent structures and fewer LOC.

I probably won't have time this week, but this looks like a decent idea to implement.

**Comment by Datseris** [comment]
My two cents in this whole discussion: the way it is done in the HK/opinion spread example **is by far the simplest way to do this, both programmatically, as well as conceptually**. Thus, I 100% disagree with:

> They take something that is conceptually simple (synchronous updates) and make it kind of complicated.

Every alternative I've read in this issue discussion is significantly more complicated, and certainly much more difficult to understand for a socioeconomical modeler, who has a 0.01% probability of knowing what the combination of `@atonce` and `@pending` macros will do.

Hence, from my perspective, performance is the reason to consider the proposed alternatives: is it faster to do it with the cryptic macro way, because either (1) we store less memory (we don't need to duplicate synchronous fields) or (2) for whatever reason the macros bring some magic that make it faster. However, given the amount of times the keyword `global` appears in the code, I would only believe that after benchmarks :D
